### ISSUE TITLE
Questions about KD loss

### ISSUE THREAD

**Issue by Paper99** [yuanli2333/Teacher-free-Knowledge-Distillation#5]
Hello!
I noticed that you didn't use `batch_mean` for calculating the [KL loss](https://github.com/yuanli2333/Teacher-free-Knowledge-Distillation/blob/2a7db7e81a35a30b88e959f19716e08cbb9067f3/my_loss_function.py#L14) on Pytorch 1.2.0.
Could you please tell me why you didn't use the `batch_mean` option?

Thanks~

**Comment by s7ev3n** [comment]
I believe this is a bug the author does not realizes! I once used `nn.KLDivLoss()` without `batch_mean`, and the experiment failed when I set `alpha=1`.

**Comment by yuanli2333** [comment]
Hi, 
It is not a bug, if you use batch_mean, you should search hyper-parameters again, which should be different with the hyper-parameters we provide here.
And it is normal that the experiment failed when you set alpha=1. For most of machine learning model,  hyper-parameters will influence the results of training, including our Tf-KD and KD.

**Comment by s7ev3n** [comment]
Well, according to the definition of KD loss, I think KL calculates the distance of student and teacher distributions, thus should divide the batch size, not element-wise, the KD loss implementation is also in [RepDistiller](https://github.com/HobbitLong/RepDistiller/blob/f51418b0d1a40dd6f4e40b6530e13f1675786e88/distiller_zoo/KD.py#L16), which it divides ` y_s.shape[0]`.

**Comment by yuanli2333** [comment]
Hi, 
There are some different choice for the batch_mean, such as in [Teacher-Assistant KD](https://github.com/imirzadeh/Teacher-Assistant-Knowledge-Distillation/blob/0f8168f4e7223134361cdd411eba0ab3b1123e29/train.py#L95) or [Pytorch Implementation](https://github.com/peterliht/knowledge-distillation-pytorch/blob/e4c40132fed5a45e39a6ef7a77b15e5d389186f8/model/net.py#L110). I agree with you that it should be averaged on batch size, it will have influence on how to choose hyper-parameters.  Tf-KD will still work when doing average on batch size, and hyper-parameters (especially \alpha) will be different and we can obtain it by search.

**Comment by Paper99** [comment]
Thank you.
